File: neet_code_150/src/math_and_geometry/plus_one.rs
struct Solution {}
impl Solution {
    // AC
    pub fn plus_one(digits: Vec<i32>) -> Vec<i32> {
        let mut res = digits;
        res.reverse();
        match res.iter_mut().next() {
            Some(v) => *v += 1,
            None => panic!("empty"),
        }

        for i in 0..res.len() {
            if res[i] == 10 {
                res[i] = 0;
                if i + 1 < res.len() {
                    res[i + 1] += 1;
                } else {
                    // 最上位のけたの繰り上がりの処理
                    res.push(1);
                }
            }
        }
        res.reverse();
        res
    }

    // 配列のサイズが大きくなるとオーバーフローしてだめ
    pub fn plus_one_b(digits: Vec<i32>) -> Vec<i32> {
        let mut num = 0;
        for (idx, v) in digits.iter().rev().enumerate() {
            num += v * 10_i32.pow(idx as u32);
        }
        //println!("num: {}", num);
        num += 1;

        let mut res = vec![];
        while num > 0 {
            res.push(num % 10);
            num /= 10;
        }
        res.reverse();

        res
    }

    // AC
    // reverse()が要らない分高速
    pub fn plus_one_c(digits: Vec<i32>) -> Vec<i32> {
        let mut res = digits;
        match res.iter_mut().rev().next() {
            Some(v) => *v += 1,
            None => panic!("empyt"),
        }

        for i in (0..res.len()).rev() {
            if res[i] == 10 {
                res[i] = 0;
                if i as i32 - 1 >= 0 {
                    res[i - 1] += 1;
                } else {
                    res.insert(0, 1);
                }
            }
        }

        res
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn plus_one(mut digits: Vec<i32>) -> Vec<i32> {
        let mut carry: u8 = 1;

        digits = digits
            .into_iter()
            .rev()
            .map(|digit| {
                let val: u8 = digit as u8 + carry;
                carry = val / 10;
                (val % 10) as i32
            })
            .collect();
        digits.reverse();

        if carry == 1 {
            digits.insert(0, 1);
        }
        digits
    }
}

fn main() {
    let case_1 = vec![1, 2, 3]; // [1,2,4]
    let case_2 = vec![4, 3, 2, 1]; // [4,3,2,1]
    let case_3 = vec![9]; // [1,0]
    let case_4 = vec![9, 8, 7, 6, 5, 4, 3, 2, 1, 0]; // [9,8,7,6,5,4,3,2,1,1]
    let case_5 = vec![4, 3, 2, 1, 3, 0];
    let case_6 = vec![9, 9];

    println!("plus_one_b()");
    println!("case_1: {:?}", Solution::plus_one_b(case_1.clone()));
    println!("case_2: {:?}", Solution::plus_one_b(case_2.clone()));
    println!("case_3: {:?}", Solution::plus_one_b(case_3.clone()));
    //println!("case_4: {:?}", Solution::plus_one_b(case_4.clone()));
    println!("case_5: {:?}", Solution::plus_one_b(case_5.clone()));

    println!("plus_one()");
    println!("case_1: {:?}", Solution::plus_one(case_1.clone()));
    println!("case_2: {:?}", Solution::plus_one(case_2.clone()));
    println!("case_3: {:?}", Solution::plus_one(case_3.clone()));
    println!("case_4: {:?}", Solution::plus_one(case_4.clone()));
    println!("case_5: {:?}", Solution::plus_one(case_5.clone()));
    println!("case_6: {:?}", Solution::plus_one(case_6.clone()));

    println!("plus_one_c()");
    println!("case_1: {:?}", Solution::plus_one_c(case_1.clone()));
    println!("case_2: {:?}", Solution::plus_one_c(case_2.clone()));
    println!("case_3: {:?}", Solution::plus_one_c(case_3.clone()));
    println!("case_4: {:?}", Solution::plus_one_c(case_4.clone()));
    println!("case_5: {:?}", Solution::plus_one_c(case_5.clone()));
    println!("case_6: {:?}", Solution::plus_one_c(case_6.clone()));
}

File: neet_code_150_cpp/math_geometry/plus_one.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Solution {
    public:
        // AC
        // ただだし速度はいまいち
        std::vector<int> plus_one(const std::vector<int>& digits) {
            auto res = digits;
            std::reverse(res.begin(), res.end());
            res[0] += 1;

            for(std::size_t i = 0; i < res.size(); ++i) {
                if(res[i] == 10) {
                    res[i] = 0;
                    if(i + 1 < res.size()) {
                        res[i + 1] += 1;
                    } else {
                        res.push_back(1);
                    }
                } 
            }

            std::reverse(res.begin(), res.end());
            return res;
        }

        // reverse()を使わない実装
        // AC
        std::vector<int> plus_one_b(const std::vector<int>& digits) {
            auto res = digits;
            res.back() += 1;

            for(int i = res.size() - 1; i >= 0; --i) {
                if(res[i] == 10) {
                    res[i] = 0;
                    if(i - 1 >= 0) {
                        res[i - 1] += 1;
                    } else {
                        res.insert(res.begin(), 1);
                    }
                }
            }

            return res;
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<int> plusOne(std::vector<int>& digits) {
        for (int i = digits.size() - 1; i >= 0; i--) {
            if (digits[i] < 9) {
                digits[i]++;
                return digits;
            }
            digits[i] = 0;
        }
        
        digits[0] = 1;
        digits.push_back(0);
        return digits;
    }
};

int main(void) {
    auto case_1 = std::vector{1, 2, 3}; // [1,2,4]
    auto case_2 = std::vector{4, 3, 2, 1}; // [4,3,2,1}
    auto case_3 = std::vector{9}; // [1,0}
    auto case_4 = std::vector{9, 8, 7, 6, 5, 4, 3, 2, 1, 0}; // [9,8,7,6,5,4,3,2,1,1}
    auto case_5 = std::vector{4, 3, 2, 1, 3, 0};
    auto case_6 = std::vector{9, 9};

   Solution s_1;

   // plus_one()
   auto res_1 = s_1.plus_one(case_1);
   for(const auto& x: res_1) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_2 = s_1.plus_one(case_2);
   for(const auto& x: res_2) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_3 = s_1.plus_one(case_3);
   for(const auto& x: res_3) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_4 = s_1.plus_one(case_4);
   for(const auto& x: res_4) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_5 = s_1.plus_one(case_5);
   for(const auto& x: res_5) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_6 = s_1.plus_one(case_6);
   for(const auto& x: res_6) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   // plus_one_b()
   auto res_1b = s_1.plus_one_b(case_1);
   for(const auto& x: res_1b) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_2b = s_1.plus_one_b(case_2);
   for(const auto& x: res_2b) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_6b = s_1.plus_one_b(case_6);
   for(const auto& x: res_6b) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   // plusOne()
   SolutionAns s_1_ans;
   auto res_1_ans = s_1_ans.plusOne(case_1);
   for(const auto& x: res_1_ans) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_2_ans = s_1_ans.plusOne(case_2);
   for(const auto& x: res_2_ans) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

   auto res_6_ans = s_1_ans.plusOne(case_6);
   for(const auto& x: res_6_ans) {
       std::cout << x << " ";
   }
   std::cout << std::endl;

}
